### PR TITLE
Add Test to wdistance calculation: Issue#11

### PR DESCRIPTION
**Pull request by whatust**
Small tests to validate coherence and word position effect on similarity.
Added new JSON  to facilitate test visualization.

### CODE REVIEW

**Comment by marco-c** [comment]
Can you merge from master? This way GitHub will only show the commits that are specific to the work you are doing now.
`git pull upstream master`

**Comment by whatust** [comment]
I cannot change from where I'm merging with the editing button, should I reopen the pull request?
"git pull upstream master" just verify that my branch is up to date.

**Comment by marco-c** [comment]
> I cannot change from where I'm merging with the editing button, should I reopen the pull request?
"git pull upstream master" just verify that my branch is up to date.

Yes, the branch is up to date with master now and there are no extraneous changes.

File: crash_similarity.py [hunk 1]
```diff
@@ -212,9 +213,9 @@ def signature_similarity(model, paths, signature1, signature2):
         for doc2 in traces2:
             words2 = [word for word in preprocess(doc2) if word in model]
 
-            if words1 == words2 or frozenset([frozenset(words1), frozenset(words2)]) in already_processed:
+            if words1 == words2 or frozenset([tuple(words1), tuple(words2)]) in already_processed:
```
Comment (marco-c) [inline crash_similarity.py]: Why did you make them tuples?

File: tests/test_basic.py [hunk 2]
```diff
@@ -5,11 +5,67 @@
...
 
 
+import crash_similarity
+
+
 class BasicTest(unittest.TestCase):
 
-    def test_pass(self):
+    # Train Model to be used in all tests
+    paths = ['crashsimilarity_data/test.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-09.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-08.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-07.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-06.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-05.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-04.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-03.json.gz']
```
Comment (marco-c) [inline tests/test_basic.py]: Let's only use `test.json.gz`. And can you move it to the tests directory?

File: tests/test_basic.py [hunk 3]
```diff
@@ -5,11 +5,67 @@
 import unittest
```
Comment (marco-c) [inline tests/test_basic.py]: Create a new file with these changes, e.g. test_crash_similarity.py, and remove the changes from this file.

File: tests/test_basic.py [hunk 4]
```diff
@@ -5,11 +5,67 @@
...
 
 
+import crash_similarity
+
+
 class BasicTest(unittest.TestCase):
 
-    def test_pass(self):
+    # Train Model to be used in all tests
```
Comment (marco-c) [inline tests/test_basic.py]: You can use a setup function to perform these initializations: https://docs.python.org/2/library/unittest.html#unittest.TestCase.setUp

File: tests/test_basic.py [hunk 5]
```diff
@@ -5,11 +5,67 @@
...
+    paths = ['crashsimilarity_data/test.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-09.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-08.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-07.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-06.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-05.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-04.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-03.json.gz']
+
+    corpus = crash_similarity.read_corpus(paths)
+
+    model = crash_similarity.train_model(corpus)
+    print('Model Trained with paths:\n{}'.format(paths))
+
+    def test_train_model(self):
```
Comment (marco-c) [inline tests/test_basic.py]: This function is not testing anything, let's remove it.

File: tests/test_basic.py [hunk 6]
```diff
@@ -5,11 +5,67 @@
...
         self.assertTrue(True)
 
+    # Test if at the number of zero distances are at least the number of different reports when pared against itself
```
Comment (marco-c) [inline tests/test_basic.py]: Can you rewrite this comment to make it clearer? Maybe also show an example of the stack traces in the `mozilla::testZeroCoherence` signature.

File: tests/test_basic.py [hunk 7]
```diff
@@ -5,11 +5,67 @@
...
+    def test_zero_dist_coherence(self):
+        signature = 'mozilla::testZeroCoherence'
+        print('\nTesting Coherence with {} signature.'.format(signature))
+
+        similarities = crash_similarity.signature_similarity(self.__class__.model, self.__class__.paths, signature, signature)
+
+        error = []
```
Comment (marco-c) [inline tests/test_basic.py]: Nit: `errors`

File: crash_similarity.py (hunk 1 above)
Comment (whatust) [inline crash_similarity.py]: if they are frozen set the order in which the words appear does not matter,
Ex
function1, function2 function3 == function2, function3, functioin1

File: tests/test_basic.py (hunk 2 above)
Comment (whatust) [inline tests/test_basic.py]: would not be insufficient data to train the model?
For now I only insert 6 entries.

File: tests/test_basic.py (hunk 4 above)
Comment (whatust) [inline tests/test_basic.py]: the setUp method is called for every test, isn't better to use setUpClass()?

File: crash_similarity.py (hunk 1 above)
Comment (marco-c) [inline crash_similarity.py]: OK, makes sense!

File: tests/test_basic.py [hunk 8]
```diff
@@ -7,7 +7,7 @@
 
 class BasicTest(unittest.TestCase):
 
-    def test_pass(self):
+    def test_train_model(self):
```
Comment (marco-c) [inline tests/test_basic.py]: Can you rename this `test_pass`? Or remove the file altogether, as it was only meant as an example.

File: tests/test_basic.py (hunk 4 above)
Comment (marco-c) [inline tests/test_basic.py]: Yes ;)

File: tests/test_basic.py (hunk 2 above)
Comment (marco-c) [inline tests/test_basic.py]: It's just for the test, so it should work.

**Comment by marco-c** [comment]
Can you merge from master again? Support for running tests on Travis was just added, so after you merge from master we will see if the tests you added are passing on Travis.

**Comment by whatust** [comment]
They are not passing, we have not implemented an algorithm that considers word position on the similarity calculation.

**Comment by marco-c** [comment]
> They are not passing, we have not implemented an algorithm that considers word position on the similarity calculation.

The first one should be passing, right?
The second can be marked as an expected failure (https://docs.python.org/2/library/unittest.html#unittest.expectedFailure).

**Comment by whatust** [comment]
when you say fail, is assert returning false or if the test are able to run?

**Comment by marco-c** [comment]
> when you say fail, is assert returning false or if the test are able to run?

assert returning false

File: crash_similarity.py [hunk 9]
```diff
@@ -90,8 +90,9 @@ def train_model(corpus):
     except:
         workers = 2
 
-    model = gensim.models.doc2vec.Doc2Vec(size=100, window=8, iter=20, workers=workers)
+    model = gensim.models.doc2vec.Doc2Vec(size=100, window=8, iter=20, workers=workers, min_count=1)
```
Comment (marco-c) [inline crash_similarity.py]: We shouldn't do this in the real world scenario. What if you add more examples to test.json instead of changing this value?

**Comment by marco-c** [comment]
Thanks!

**Pull request merged**